Estas son las librerías permitidas. Si quieren utilizar alguna librería adicional, pueden realizar la consulta a través de Discord.

In [17]:
import re
import numpy as np
import pandas as pd


**Ejercicio 1 - *Bag of Words* 🐶🐈(0.5 puntos).**



Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:

**disclaimer: El orden de los resultados pueden variar**

In [4]:
d0 = 'El perro se come la comida y después se duerme'
d1 = 'El perro se despierta y después empieza a ladrar'
d2 = 'El perro ladra y después se come la comida'
d3 = 'El gato se come la comida y después se duerme'
d4 = 'El gato se despierta y después empieza a maullar'
d5 = 'El gato maulla y después se come la comida'
dataset = [d0, d1, d2, d3, d4, d5]

El objetivo de este ejercicio es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica TF-IDF.

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es el **Bag of Words**, método mediante el cuál se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que reciba como input un arreglo de documentos y devuelva un pandas dataframe con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representa las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el bow de un determinado documento.


Por ejemplo para el siguiente dataset:

```
dataset = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |



In [10]:
import nltk
nltk.download('punkt')
from nltk import word_tokenize

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


Podría serle útil primero definir una funcion `get_vocab` y `tokenize`.

In [16]:
def bag_of_words(dataset):
  ### Aquí inicia tu código ###
  pass
  ### Aquí termina tu código ###

***Test:***

In [ ]:
dataset_bow = bag_of_words(dataset)
dataset_bow

``
``

***Resultado esperado***:

|    | El | perro | se | come | la | comida | y | después | duerme | despierta | empieza | a | ladrar | ladra | gato | maullar | maulla |
|----|---:|------:|---:|-----:|---:|-------:|--:|--------:|-------:|----------:|--------:|--:|-------:|------:|-----:|--------:|-------:|
| d0 |  1 |     1 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    0 |       0 |      0 |
| d1 |  1 |     1 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      1 |     0 |    0 |       0 |      0 |
| d2 |  1 |     1 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     1 |    0 |       0 |      0 |
| d3 |  1 |     0 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      0 |
| d4 |  1 |     0 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      0 |     0 |    1 |       1 |      0 |
| d5 |  1 |     0 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      1 |

``
``

` `  
` `

**Ejercicio 2 - *Calcular TF* (0.5 puntos):** Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia ${max_i({\text{tf}_{i,j})}}$, donde
i corresponde al índice de las filas (bow) y j al de las columnas (palabras). Es decir, dividir cada bow en la cantidad de veces de la palabra que aparezca más veces en ese vector.


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{max_i({\text{tf}_{i,j})}}$$

In [ ]:
def calc_tf(dataset_bow):
    ### Aquí inicia tu código ###
    pass
    ### Aquí termina tu código ###

***Test:***

In [ ]:
tf = calc_tf(dataset_bow)
tf

a  comida   El   se    y  maullar  come  ladra  despierta  ladrar  \
d0  0.0     0.5  0.5  1.0  0.5      0.0   0.5    0.0        0.0     0.0   
d1  1.0     0.0  1.0  1.0  1.0      0.0   0.0    0.0        1.0     1.0   
d2  0.0     1.0  1.0  1.0  1.0      0.0   1.0    1.0        0.0     0.0   
d3  0.0     0.5  0.5  1.0  0.5      0.0   0.5    0.0        0.0     0.0   
d4  1.0     0.0  1.0  1.0  1.0      1.0   0.0    0.0        1.0     0.0   
d5  0.0     1.0  1.0  1.0  1.0      0.0   1.0    0.0        0.0     0.0   

    maulla  gato  después   la  perro  empieza  duerme  
d0     0.0   0.0      0.5  0.5    0.5      0.0     0.5  
d1     0.0   0.0      1.0  0.0    1.0      1.0     0.0  
d2     0.0   0.0      1.0  1.0    1.0      0.0     0.0  
d3     0.0   0.5      0.5  0.5    0.0      0.0     0.5  
d4     0.0   1.0      1.0  0.0    0.0      1.0     0.0  
d5     1.0   1.0      1.0  1.0    0.0      0.0     0.0

``
``

***Resultado esperado***:

|    |  El | perro |  se | come |  la | comida |   y | después | duerme | despierta | empieza |   a | ladrar | ladra | gato | maullar | maulla |
|----|----:|------:|----:|-----:|----:|-------:|----:|--------:|-------:|----------:|--------:|----:|-------:|------:|-----:|--------:|-------:|
| d0 | 0.5 |   0.5 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d1 | 1.0 |   1.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    1.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d2 | 1.0 |   1.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   1.0 |  0.0 |     0.0 |    0.0 |
| d3 | 0.5 |   0.0 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.5 |     0.0 |    0.0 |
| d4 | 1.0 |   0.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    0.0 |   0.0 |  1.0 |     1.0 |    0.0 |
| d5 | 1.0 |   0.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  1.0 |     0.0 |    1.0 |

``
``

` `  
` `

**Ejercicio 3 - *Calcular IDF* (0.5 punto):**


Implementar `calc_idf(dataset_bow)`. Este debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el calculo de cada idf por palabra.

Recordar que $idf_{t_i} = log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ Número de documentos que contienen la palabra $t_i$

In [ ]:
def calc_idf(dataset_bow):
    ### Aquí inicia tu código ###
    pass
    ### Aquí termina tu código ###

***Test:***

In [ ]:
idf = calc_idf(dataset_bow)
idf

{'a': 0.47712125471966244,
 'comida': 0.17609125905568124,
 'El': 0.0,
 'se': 0.0,
 'y': 0.0,
 'maullar': 0.7781512503836436,
 'come': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'despierta': 0.47712125471966244,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'gato': 0.3010299956639812,
 'después': 0.0,
 'la': 0.17609125905568124,
 'perro': 0.3010299956639812,
 'empieza': 0.47712125471966244,
 'duerme': 0.47712125471966244}

***Resultado esperado***:

```python
{'El': 0.0,
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}
```

Puede notar el bajo puntaje otorgado a las palabras que más se repiten! 😮

**Ejercicio 4 - *Calcular TF-IDF & concluir similitud de documentos.* (1 punto)**


Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [ ]:
def calc_tf_idf(tf, idf):
    ### Aquí inicia tu código ###

    pass

    ### Aquí termina tu código ###

***Test.***

In [ ]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

a    comida   El   se    y   maullar      come     ladra  \
d0  0.000000  0.088046  0.0  0.0  0.0  0.000000  0.088046  0.000000   
d1  0.477121  0.000000  0.0  0.0  0.0  0.000000  0.000000  0.000000   
d2  0.000000  0.176091  0.0  0.0  0.0  0.000000  0.176091  0.778151   
d3  0.000000  0.088046  0.0  0.0  0.0  0.000000  0.088046  0.000000   
d4  0.477121  0.000000  0.0  0.0  0.0  0.778151  0.000000  0.000000   
d5  0.000000  0.176091  0.0  0.0  0.0  0.000000  0.176091  0.000000   

    despierta    ladrar    maulla      gato  después        la     perro  \
d0   0.000000  0.000000  0.000000  0.000000      0.0  0.088046  0.150515   
d1   0.477121  0.778151  0.000000  0.000000      0.0  0.000000  0.301030   
d2   0.000000  0.000000  0.000000  0.000000      0.0  0.176091  0.301030   
d3   0.000000  0.000000  0.000000  0.150515      0.0  0.088046  0.000000   
d4   0.477121  0.000000  0.000000  0.301030      0.0  0.000000  0.000000   
d5   0.000000  0.000000  0.778151  0.301030      0.0  0.176091  0.000000   

     empieza    duerme  
d0  0.000000  0.238561  
d1  0.477121  0.000000  
d2  0.000000  0.000000  
d3  0.000000  0.238561  
d4  0.477121  0.000000  
d5  0.000000  0.000000

``
``

***Resultado esperado***:

|    |  El |    perro |  se |     come |       la |   comida |   y | después |   duerme | despierta |  empieza |        a |   ladrar |    ladra |     gato |  maullar |   maulla |
|----|----:|---------:|----:|---------:|---------:|---------:|----:|--------:|---------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|
| d0 | 0.0 | 0.150515 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d1 | 0.0 | 0.301030 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.778151 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d2 | 0.0 | 0.301030 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.778151 | 0.000000 | 0.000000 | 0.000000 |
| d3 | 0.0 | 0.000000 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.150515 | 0.000000 | 0.000000 |
| d4 | 0.0 | 0.000000 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.000000 | 0.000000 | 0.301030 | 0.778151 | 0.000000 |
| d5 | 0.0 | 0.000000 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.301030 | 0.000000 | 0.778151 |


``
``

Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica. Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos vectores. Concluya cuáles son los dos documentos más similares.

In [ ]:
def cosine_similarity(v1, v2):
  ### Aquí inicia tu código ###

  pass

  ### Aquí termina tu código ###

similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity


![gato](https://live.staticflickr.com/4652/38904147065_0b6c446945_b.jpg)

**Cualquier recomendación que nos quisieran dar para una futura tarea es bienvenid@!**